File: frontend/public/apps/sanction/scripts/update_data.js
const axios = require('axios');
const xml2js = require('xml2js');
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');

const DATA_FILE = path.join(__dirname, '../sanctions_data.json');

// Sources
const SOURCES = {
    UN: 'https://scsanctions.un.org/resources/xml/en/consolidated.xml',
    EU: 'https://webgate.ec.europa.eu/europeaid/fsd/fsf/public/files/xmlFullSanctionsList/content?token=dG9rZW4tMjAxNw',
    OFAC: 'https://www.treasury.gov/ofac/downloads/sdn.csv'
};

let consolidatedData = [];

async function fetchData() {
    console.log('Starting data update...');

    try {
        await fetchUN();
    } catch (e) { console.error('Error fetching UN:', e.message); }

    try {
        await fetchEU();
    } catch (e) { console.error('Error fetching EU:', e.message); }

    try {
        await fetchOFAC();
    } catch (e) { console.error('Error fetching OFAC:', e.message); }

    // Save to file
    fs.writeFileSync(DATA_FILE, JSON.stringify(consolidatedData, null, 2));
    console.log(`Data update complete. Total records: ${consolidatedData.length}`);
}

async function fetchUN() {
    console.log('Fetching UN List...');
    const response = await axios.get(SOURCES.UN);
    const parser = new xml2js.Parser();
    const result = await parser.parseStringPromise(response.data);

    const individuals = result.CONSOLIDATED_LIST.INDIVIDUALS[0].INDIVIDUAL || [];
    const entities = result.CONSOLIDATED_LIST.ENTITIES[0].ENTITY || [];

    individuals.forEach(item => {
        const name = `${item.FIRST_NAME} ${item.SECOND_NAME || ''} ${item.THIRD_NAME || ''}`.trim();
        consolidatedData.push({
            name: name,
            list: 'UN Consolidated',
            type: 'Individual',
            originalId: item.DATAID[0]
        });
    });

    entities.forEach(item => {
        consolidatedData.push({
            name: item.FIRST_NAME[0],
            list: 'UN Consolidated',
            type: 'Entity',
            originalId: item.DATAID[0]
        });
    });
    console.log(`Parsed ${individuals.length + entities.length} UN records.`);
}

async function fetchEU() {
    console.log('Fetching EU List...');
    const response = await axios.get(SOURCES.EU);
    const parser = new xml2js.Parser();
    const result = await parser.parseStringPromise(response.data);

    const subjects = result.export.sanctionEntity || [];

    subjects.forEach(item => {
        // EU structure is complex, simplified for MVP
        const nameAlias = item.nameAlias ? item.nameAlias[0].wholeName[0] : 'Unknown';
        consolidatedData.push({
            name: nameAlias,
            list: 'EU Financial Sanctions',
            type: 'Entity/Individual',
            originalId: item.$.logicalId
        });
    });
    console.log(`Parsed ${subjects.length} EU records.`);
}

async function fetchOFAC() {
    console.log('Fetching OFAC List...');
    const response = await axios.get(SOURCES.OFAC, { responseType: 'stream' });

    return new Promise((resolve, reject) => {
        let count = 0;
        response.data.pipe(csv(['ent_num', 'SDN_Name', 'SDN_Type', 'Program', 'Title', 'Call_Sign', 'Vess_type', 'Tonnage', 'GRT', 'Vess_flag', 'Vess_owner', 'Remarks']))
            .on('data', (row) => {
                if (row.SDN_Name) {
                    consolidatedData.push({
                        name: row.SDN_Name,
                        list: 'OFAC SDN',
                        type: row.SDN_Type,
                        originalId: row.ent_num
                    });
                    count++;
                }
            })
            .on('end', () => {
                console.log(`Parsed ${count} OFAC records.`);
                resolve();
            })
            .on('error', reject);
    });
}

fetchData();

File: frontend/public/apps/sanction/scripts/update_data.py
import urllib.request
import xml.etree.ElementTree as ET
import csv
import json
import ssl
import sys
import io
import os

# Disable SSL verification for simplicity (avoid cert errors)
ssl._create_default_https_context = ssl._create_unverified_context

# Save to parent directory (project root)
DATA_FILE = os.path.join(os.path.dirname(__file__), '..', 'sanctions_data.js')

SOURCES = {
    'UN': 'https://scsanctions.un.org/resources/xml/en/consolidated.xml',
    'EU': 'https://webgate.ec.europa.eu/europeaid/fsd/fsf/public/files/xmlFullSanctionsList/content?token=dG9rZW4tMjAxNw',
    'OFAC': 'https://www.treasury.gov/ofac/downloads/sdn.csv'
}

consolidated_data = []

def fetch_un():
    print("Fetching UN List...")
    try:
        req = urllib.request.Request(SOURCES['UN'], headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            xml_content = response.read()
            root = ET.fromstring(xml_content)
            
            # Individuals
            for individual in root.findall('.//INDIVIDUAL'):
                first_name = individual.find('FIRST_NAME').text if individual.find('FIRST_NAME') is not None else ''
                second_name = individual.find('SECOND_NAME').text if individual.find('SECOND_NAME') is not None else ''
                third_name = individual.find('THIRD_NAME').text if individual.find('THIRD_NAME') is not None else ''
                name = f"{first_name} {second_name} {third_name}".strip()
                
                consolidated_data.append({
                    'name': name,
                    'list': 'UN Consolidated',
                    'type': 'Individual',
                    'originalId': individual.find('DATAID').text
                })

            # Entities
            for entity in root.findall('.//ENTITY'):
                name = entity.find('FIRST_NAME').text if entity.find('FIRST_NAME') is not None else ''
                consolidated_data.append({
                    'name': name,
                    'list': 'UN Consolidated',
                    'type': 'Entity',
                    'originalId': entity.find('DATAID').text
                })
            print(f"Parsed UN records.")
    except Exception as e:
        print(f"Error fetching UN: {e}")

def fetch_eu():
    print("Fetching EU List...")
    try:
        req = urllib.request.Request(SOURCES['EU'], headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            xml_content = response.read()
            root = ET.fromstring(xml_content)
            
            count = 0
            for entity in root.findall('.//sanctionEntity'):
                name_alias = entity.find('.//nameAlias')
                if name_alias is not None:
                    name = name_alias.find('wholeName').text
                    consolidated_data.append({
                        'name': name,
                        'list': 'EU Financial Sanctions',
                        'type': 'Entity/Individual',
                        'originalId': entity.get('logicalId')
                    })
                    count += 1
            print(f"Parsed {count} EU records.")
    except Exception as e:
        print(f"Error fetching EU: {e}")

def fetch_ofac():
    print("Fetching OFAC List...")
    try:
        req = urllib.request.Request(SOURCES['OFAC'], headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            csv_content = response.read().decode('utf-8')
            reader = csv.reader(io.StringIO(csv_content))
            
            count = 0
            for row in reader:
                if len(row) > 1:
                    # OFAC CSV format: ent_num, SDN_Name, SDN_Type, ...
                    consolidated_data.append({
                        'name': row[1], # SDN_Name
                        'list': 'OFAC SDN',
                        'type': row[2], # SDN_Type
                        'originalId': row[0]
                    })
                    count += 1
            print(f"Parsed {count} OFAC records.")
    except Exception as e:
        print(f"Error fetching OFAC: {e}")

def main():
    fetch_un()
    fetch_eu()
    fetch_ofac()
    
    # Write as JS file
    with open(DATA_FILE, 'w', encoding='utf-8') as f:
        json_str = json.dumps(consolidated_data, indent=2, ensure_ascii=False)
        f.write(f"window.SANCTIONS_DATA = {json_str};")
    
    print(f"Total records saved to {DATA_FILE}: {len(consolidated_data)}")

if __name__ == "__main__":
    main()
